### DISCUSSION TITLE
LUT cells get ignored/deleted by Vivado

### DISCUSSION THREAD

**Discussion by codefreak472** [Xilinx/RapidWright#634] [Q&A] [answered]
Hello, 

I'm trying to get a feel for the basic functions of rapidwright and I'm already having some issues. If I use the design.createAndPlaceCell method to create a LUT cell, I can see that it's been created just fine in my design object. However, when I write that design to a file, and open it in Vivado, the LUT cell is gone. I have tried LUTs with and without init strings, as well as with the location locked, and the DONT_TOUCH property set. None of them have helped. My design is simply empty when I open it in Vivado.

**Comment by eddieh-xlnx** [comment]
Welcome @codefreak472!

Nothing comes to mind as to why this wouldn't work. A really simple example is here: https://github.com/Xilinx/RapidWright/blob/f5930b27f0c48918581274ef42f71f2fd161a25c/src/com/xilinx/rapidwright/examples/Lesson1.java

You can also find some other examples here: https://github.com/Xilinx/RapidWright/search?q=createAndPlaceCell

Failing those, can you share your code so that we may reproduce it on our end?

**Reply by codefreak472** [reply]
I was able to recreate my problem with the HelloWorld design. Copying all the code for that example runs without issue. However, if I try saving off a checkpoint after cell creation/placement, but before net creation/connection, opening that dcp results in an empty design with several warning messages saying my instances don't exist. The graphic in the RapidWright introduction page (http://www.rapidwright.io/docs/Introduction.html) led me to believe I could save off a placed design like this. I am very new at this so maybe I'm not understanding what it means to have properly placed design?

**Reply by eddieh-xlnx** [reply]
You're right. I can reproduce what you're seeing. It looks like, by default, Vivado is optimizing away unconnected `Cell`-s.

You mentioned that:
> I have tried LUTs with and without init strings, as well as with the location locked, and the DONT_TOUCH property set.

How are you setting this `DONT_TOUCH` property? I'm doing it like this:
```
--- a/src/com/xilinx/rapidwright/examples/Lesson1.java
+++ b/src/com/xilinx/rapidwright/examples/Lesson1.java
@@ -55,26 +55,9 @@ public class Lesson1 {
         Cell button1 = d.createAndPlaceIOB("button1", PinType.IN , "D20",  "LVCMOS33");
         Cell led0    = d.createAndPlaceIOB("led0"   , PinType.OUT, "R14",  "LVCMOS33");
 
-        // Connect Button 0 to the LUT2 input I0
-        Net net0 = d.createNet("button0_IBUF");
-        net0.connect(button0, "O");
-        net0.connect(and2, "I0");
-
-        // Connect Button 1 to the LUT2 input I1
-        Net net1 = d.createNet("button1_IBUF");
-        net1.connect(button1, "O");
-        net1.connect(and2, "I1");
-
-        // Connect the LUT2 (AND2) to the LED IO
-        Net net2 = d.createNet("and2");
-        net2.connect(and2, "O");
-        net2.connect(led0, "I");
-
-        // Route site internal nets
-        d.routeSites();
-
-        // Route nets between sites
-        new Router(d).routeDesign();
+        for (Cell c : d.getCells()) {
+            c.getEDIFCellInst().addProperty("DONT_TOUCH", true);
+        }
 
         // Save our work in a Checkpoint
         d.writeCheckpoint("HelloWorld.dcp");
```
and it looks to be working now.

**Reply by codefreak472** [reply]
Interestingly, that half solved the problem for me. My LUTs are no longer disappearing, but they are unplaced when I open the design. 

I was using cell.setPropertiesMap() before. what's the difference between using cell.getEDIFCellInst().addProperty() and setPropertiesMap()?

**Reply by codefreak472** [reply]
doing the same thing as before with the hello world example, if I set the DONT_TOUCH property for the cells, and save a dcp before doing any routing, when I open the dcp, the IOBs are still created and placed, but the LUT is unplaced. I don't understand why it's only LUTs that seem to have this issue.

**Reply by eddieh-xlnx** [reply]
Sorry I didn't notice that the LUT was unplaced. I've looked at it some more, I'm not quite sure what the reason for this behaviour is. Since it's very unconventional to want to have a completely unconnected LUT in your design, it's something we haven't seen before. In fact, even after manually placing the LUT using Vivado, writing out the DCP and then loading it back, Vivado itself raises the following warning:
```
CRITICAL WARNING: [Designutils 20-756] Invalid physical equation for the A6LUT bel in site SLICE_X100Y100. The original INIT is '8'. The logical cell is 'and2'. Reason: The bit width of the INIT value does not match the number of used input pins '0'. Please verify that all required logical pins are used for this cell.
     Bel 14: element name: 'A6LUT'
      Attr: 'LUT_OR_MEM' Value: LUT
      Attr: 'RAMMODE' Value: NOT CONFIGURED
      Attr: 'EQN' Value: 4'h8
      Inst: 'and2'
      CellType: 'LUT2'
      Pin Swappable:
        Not assigned -> Bel pin 'A1'
        Not assigned -> Bel pin 'A2'
        Not assigned -> Bel pin 'A3'
        Not assigned -> Bel pin 'A4'
        Not assigned -> Bel pin 'A5'
        Not assigned -> Bel pin 'A6'
```

Please leave it with us to investigate.

Can you share the reason you need this floating LUT inside Vivado? If this isn't just a stepping stone to something bigger, perhaps there's another way of achieving what you need.

**Reply by codefreak472** [reply]
Since I am very new at this, I thought I could save my work at multiple points throughout the process to verify I was doing everything correctly. Since the LUT is disappearing, I was assuming that meant I was doing something wrong during placement. The goal is not to have a floating LUT in the design, it is meant purely as a stepping stone. It still strikes me as odd that the placement is lost, even if the DONT_TOUCH property is set though. I thought I could create the LUT with RapidWright, then open the placed design in Vivado and do routing in Vivado, but this doesn't seem possible.

**Reply by codefreak472** [reply]
After some additional toying with the HelloWorld design, if I save the design with design.writeCheckpoint() after connecting net0 to the LUT, the LUT appears in the design and is placed. So it looks like connecting a net to a cell is sufficient to convince Vivado not to get rid of it/unplace it.

I have had a lot of trouble with the net.connect() method as well though. I've gotten a lot of NullPointerException errors from it. I think it happens if a SitePinInst already exists for the pin you want to connect, but I'm not sure.

**Reply by eddieh-xlnx** [reply, marked as answer]
> Since I am very new at this, I thought I could save my work at multiple points throughout the process to verify I was doing everything correctly. Since the LUT is disappearing, I was assuming that meant I was doing something wrong during placement. The goal is not to have a floating LUT in the design, it is meant purely as a stepping stone.

That is absolutely the right idea, and something that RapidWright does support, please don't be dissuaded from this approach!

> I thought I could create the LUT with RapidWright, then open the placed design in Vivado and do routing in Vivado, but this doesn't seem possible.

That is absolutely possible (as you can see in the `HelloWorld` example) for the much more common use case where the LUT is actually connected to something. In your case, you're inserting a floating LUT and in doing so stumbled on a bug we've yet to track down.

The stepping stone you're actually trying to achieve is actually missing two steps before you can get to a meaningful circuit:
1. Connecting the LUT and I/Os to each other
2. Routing those connections (which you've mentioned)

Both of which you'll need to do with Vivado if you just place a floating LUT in the design. If you were to perform 1 with RapidWright, then that should bring you back onto the well-tested path for now.

**Reply by eddieh-xlnx** [reply]
> I have had a lot of trouble with the net.connect() method as well though. I've gotten a lot of NullPointerException errors from it. I think it happens if a SitePinInst already exists for the pin you want to connect, but I'm not sure.

Please share these issues. We absolutely want to make RapidWright easier to use, either by enhancing the codebase or through better documentation!
